### ISSUE TITLE
re-add `AddProperty`

### ISSUE THREAD

**Issue by Rerago** [LeagueToolkit/LeagueToolkit#244]
In the refactor, the function `AddProperty` has been removed. Is there a reason for that?

It does not bother. But there is a duplication when using the `Add` of the `Dictionary`. Because the `NameHash` must already be specified when creating the added object.

The following code

```cs
var newAnchors = new BinTreeStructure(positionObj, AnchorsNameHash, AnchorSingleNameHash, new List<BinTreeProperty>());
newAnchors.AddProperty(new BinTreeVector2(newAnchors, AnchorNameHash, new Vector2(0, 0)));
positionObj.AddProperty(newAnchors);
```
would become the following code after the refactor

```cs
var newAnchors = new BinTreeStruct(AnchorsNameHash, AnchorSingleNameHash, new List<BinTreeProperty>());
newAnchors.Properties.Add(AnchorNameHash, new BinTreeVector2(AnchorNameHash, new Vector2(0, 0)));
positionObj.Properties.Add(AnchorsNameHash, newAnchors);
```
And as you can see I had to specify `AnchorsNameHash` and `AnchorNameHash` twice. Whereas once should be enough when creating.

PS: why is it not necessary to specify a parent for the `BinTreeStructure` anymore (or why was it necessary before?).

**Comment by Crauzer** [comment]
I see where you're coming from. I removed all of the "helper" methods because I refactored the lists into dictionaries so they are faster and easier to index. Although as you mention, this does create some redundancy as shown above.

The best approach here would be to create an extension method for ´IDictionary<uint, BinTreeProperty>´ which accepts only the property as a parameter and adds it to the dictionary using the name hash as the key.

**Comment by Crauzer** [comment]
> PS: why is it not necessary to specify a parent for the BinTreeStructure anymore (or why was it necessary before?).

The IBinTreeParent was useless, it didn't actually do anything, so I removed it, it also doesn't really make sense to have it since there isn't any behavior which would require it.

**Comment by Rerago** [comment]
All right, perfect. If you want to implement it directly, I have a suggestion:

```cs
public static class BinTreePropertyDictExtensions
{
    public static void Add(this IDictionary<uint, BinTreeProperty> obj, BinTreeProperty value)
        => obj.Add(value.NameHash, value);

    public static bool TryAdd(this IDictionary<uint, BinTreeProperty> obj, BinTreeProperty value)
        => obj.TryAdd(value.NameHash, value);

    public static T GetValueOrDefault<T>(this IDictionary<uint, BinTreeProperty> obj, uint nameHash) where T : BinTreeProperty
        => obj.TryGetValue<T>(nameHash, out var value) == true ? value : default;

    public static bool TryGetValue<T>(this IDictionary<uint, BinTreeProperty> obj, uint nameHash, out T value) where T : BinTreeProperty
    {
        var returnVal = obj.TryGetValue(nameHash, out BinTreeProperty prop) == true;
        value = returnVal && prop is T propTyped ? propTyped : null;
        return returnVal;
    }
}
```
- `Add` and `TryAdd` are simply overloads for the default functions of the `IDictionary`
- `GetValueOrDefault` and `TryGetValue` are quite useful to get the property in the correct type

**Comment by Crauzer** [comment]
Could you explain how `GetValueOrDefault` is useful ?
The `default` value of all reference types in C# is `null` and the value returned by `TryGetValue` is `default` if the result is `false`

**Comment by Crauzer** [comment]
Code:
```cs
public static class BinTreePropertyDictionaryExtensions
{
    public static void Add(this IDictionary<uint, BinTreeProperty> dictionary, BinTreeProperty property) =>
        dictionary.Add(property.NameHash, property);

    public static bool TryAdd(this IDictionary<uint, BinTreeProperty> dictionary, BinTreeProperty property) =>
        dictionary.TryAdd(property.NameHash, property);

    public static bool TryGetValue<T>(this IDictionary<uint, BinTreeProperty> dictionary, uint nameHash, out T property)
        where T : BinTreeProperty
    {
        bool result = dictionary.TryGetValue(nameHash, out BinTreeProperty value);
        property = value switch
        {
            T concreteProperty => concreteProperty,
            _ => default
        };

        return result;
    }
}
```

**Comment by Rerago** [comment]
About the naming:

- I took `BinTreeProperty value` because MS used the property `value` in `Add` and `TryAdd`
- GetValueOrDefault' I added because of [system.collections.generic.collectionextensions.getvalueordefault](https://learn.microsoft.com/dotnet/api/system.collections.generic.collectionextensions.getvalueordefault)

I am aware that `default` of all reference classes is always null. But I just wanted to stick to the default function. Of course I could also have written the following:

```cs
=> obj.TryGetValue<T>(nameHash, out var value) == true ? value : null;
```

Surely it could have been called just `GetValue`. But I wanted to stick to the function name of the interface. The advantage of a direct function without `Try` is that the code can be shorter.

**Comment by Rerago** [comment]
About the naming:

- I took `BinTreeProperty value` because MS used the property `value` in `Add` and `TryAdd`
- `GetValueOrDefault` I added because of [system.collections.generic.collectionextensions.getvalueordefault](https://learn.microsoft.com/dotnet/api/system.collections.generic.collectionextensions.getvalueordefault)

I am aware that `default` of all reference classes is always `null`. But I just wanted to stick to the default function. Of course I could also have written the following:

```cs
=> obj.TryGetValue<T>(nameHash, out var value) == true ? value : null;
```

Surely it could have been called just `GetValue`. But I wanted to stick to the function name of the interface. The advantage of a direct function without `Try` is that the code can be shorter.
